picoquic/picoquic_unified_log.h

```diff
@@ -0,0 +1,189 @@
+
+/* Log arrival or departure of an UDP datagram for an unknown connection */
+typedef void (*picoquic_log_quic_pdu_fn)(picoquic_quic_t* quic, int receiving, uint64_t current_time, uint64_t cid64,
+    const struct sockaddr* addr_peer, const struct sockaddr* addr_local, size_t packet_length);
+
+/* Log an event relating to a specific connection */
+typedef void (*picoquic_log_app_message_fn)(picoquic_cnx_t* cnx, const char* fmt, va_list vargs);
+
+/* Log arrival or departure of an UDP datagram on a connection */
+typedef void (*picoquic_log_pdu_fn)(picoquic_cnx_t* cnx, int receiving, uint64_t current_time,
+    const struct sockaddr* addr_peer, const struct sockaddr* addr_local, size_t packet_length);
+
+/* Log a decrypted packet - receiving = 1 if arrival, = 0 if sending */
+typedef void (*picoquic_log_packet_fn)(picoquic_cnx_t* cnx, int receiving, uint64_t current_time,
+    struct st_picoquic_packet_header_t* ph, const uint8_t* bytes, size_t bytes_max);
+
+/* Report that a packet was dropped due to some error */
+typedef void (*picoquic_log_dropped_packet_fn)(picoquic_cnx_t* cnx, struct st_picoquic_packet_header_t* ph, size_t packet_size, int err, uint8_t* raw_data, uint64_t current_time);
+
+/* Report that packet was buffered waiting for decryption */
+typedef void (*picoquic_log_buffered_packet_fn)(picoquic_cnx_t* cnx, picoquic_packet_type_enum ptype, uint64_t current_time);
+
+/* Log that a packet was formatted, ready to be sent. */
+typedef void (*picoquic_log_outgoing_packet_fn)(picoquic_cnx_t* cnx,
+    uint8_t* bytes, uint64_t sequence_number, size_t pn_length, size_t length,
+    uint8_t* send_buffer, size_t send_length, uint64_t current_time);
+
+/* Log packet lost events */
+typedef void (*picoquic_log_packet_lost_fn)(picoquic_cnx_t* cnx,
+    picoquic_packet_type_enum ptype, uint64_t sequence_number, char const* trigger,
+    picoquic_connection_id_t* dcid, size_t packet_size,
+    uint64_t current_time);
+
+/* log negotiated ALPN */
+typedef void (*picoquic_log_negotiated_alpn_fn)(picoquic_cnx_t* cnx, int is_local,
+    uint8_t const* sni, size_t sni_len, uint8_t const* alpn, size_t alpn_len,
+    const ptls_iovec_t* alpn_list, size_t alpn_count);
+
+/* log transport extension -- either formatted by the loacl peer (is_local=1) or received from remote peer */
+typedef void (*picoquic_log_transport_extension_fn)(picoquic_cnx_t* cnx, int is_local,
+    size_t param_length, uint8_t* params);
+
+/* log TLS ticket */
+typedef void (*picoquic_log_tls_ticket_fn)(picoquic_cnx_t * cnx,
+    uint8_t* ticket, uint16_t ticket_length);
+
+/* log the start of a connection */
+typedef void (*picoquic_log_new_connection_fn)(picoquic_cnx_t* cnx);
+/* log the end of a connection */
+typedef void (*picoquic_log_close_connection_fn)(picoquic_cnx_t* cnx);
+
+/* log congestion control parameters */
+typedef void (*picoquic_log_cc_dump_fn)(picoquic_cnx_t* cnx, uint64_t current_time);
+
+typedef struct st_picoquic_unified_login_t {
```

Comment:
Should probably read `st_picoquic_unified_logging_t` :)

Reply:
Yes! Fixing that too.